Centralized progress dialog

antis81 opened the pull request:
Provide a central place to visualize progress of running operations.
- [x] **Part 1**
  - [x] Copy existing `ProgressDlg` from "Repository" module to `libMacGitverCore`
    - Side note: The dialog was used in the "CloneDialog".
  - [x] Make it generic :arrow_right: redefine slots to handle reported progress)
  - [x] Add concept to add an activity, consisting of multiple steps.
  - [x] Make it look good. :star2:
- [ ] **Part 2**
  - [x] Create a `libActivitiesGui`.
  - [x] Copy ProgressDlg to `libActivitiesGui`.
  - [ ] Delete ProgressDlg in `libMacGitverCore`.
  - [ ] Make it a single instance, managed by `Activity::Manage??`.
  - [ ] Reimplement & rewire to Activity framework.
  - [ ] Rethink the logging: Each activity (service?) has it's own log.

scunz committed:
Move some RepoMan code around

scunz committed:
Make AutoRefresher internal

scunz committed:
Remove signals from RM::Repo

scunz committed:
Change RepoMan's "Set" to "List" everywhere

scunz committed:
Make the RM::Repo's Git::Repository member private

scunz committed:
Split off RepoMan & make everything compile at least

scunz committed:
Move refresher instance into MacGitver

scunz committed:
Repoman Events

scunz committed:
Fix some Compat-Event compile issues

antis81 committed:
add ProgressWdgt, preparing for unified progress dialog

The ProgressWdgt is going to be used internally by ProgressDlg to represent a single activity or step.

scunz committed:
Logging Channel for Git

scunz committed:
Temporary code

antis81 committed:
copied progress dialog from "Repository" module (ProgressDlg) to libMacGitverCore with required minor changes

antis81 committed:
redesign ProgressDlg to visualize the progress of multiple activities

An activity must contain at least one step to be processed.

Note: There's no logging available yet.

scunz commented on Libs/libMacGitverCore/Widgets/ProgressDlg.cpp:
```diff
@@ -0,0 +1,206 @@
+
+    public:
+        Steps   mSteps;
+        Status  mStatus     = Running;
```
Please don't do this, unless it's a _static_ initialiser. Confusion in constructors as what happens when is too error-prone otherwise.

antis81 replied:
Hm, okay ... You mean like error-prone concerning code readability or technically?

scunz replied:
Both, actually. If I found a constructor only initialising half of the members, I'd probably consider this an error and fix it - though there might not be a problem at all, considering that some member could be initialised in the class definition.

Technically spoken: At which point can you assume which variable to be initialised? We know the order changed from declaration-order to access-then-declaration-order from C++03 to C++11. Trouble starts when you have to use `this` inside a constructor's base-class or member-initialiser-list...

scunz commented on Libs/libMacGitverCore/Widgets/ProgressDlg.cpp:
```diff
@@ -0,0 +1,206 @@
+{
+    Q_ASSERT(activity);
+
+    Private::ProgressWdgt* a( new Private::ProgressWdgt(description) );
```
Using the Copy-CTor of a pointer seems wired at least. I'd prefer to see an assignment here (at least to show clearly what happens - You know that C++ has strange rules on calling move/copy/assignment ctor/operator regardless if the codes uses a ctor or assignment)...

antis81 commented:
@scunz Btw.: Could you please fix the following compiler error in `development` branch (just a little one) and make it compile again? I can rebase this PR afterwards.

``` bash
[ 13%] Building CXX object Libs/libMacGitverCore/CMakeFiles/MacGitverCore.dir/RepoMan/AutoRefresher.cpp.o                       
/home/nils/Projects/MacGitver/Libs/libMacGitverCore/RepoMan/AutoRefresher.cpp:196:9: error: no matching function for call to    
      'log'
        MacGitver::log().addMessage(trUtf8("Refreshing git repositories..."));
        ^~~~~~~~~~~~~~
/home/nils/Projects/MacGitver/Libs/libMacGitverCore/App/MacGitver.hpp:66:17: note: candidate function not viable: requires 2    
      arguments, but 0 were provided
    static void log( Log::Type type, const QString& logMessage );
```

Thanks!

scunz commented:
So, I've been reviewing this code now 3 or 4 times - and this time I've left some traces of it :)

However, when I first read it, I was worried about one thing: Who's going to synchronise this code - I've thought for quite a time that it's possible to do that in the `ActivityManager`, but reading it over and over again and thinking about how I'm going to do that `ActivityManager` thing, I've come to the conclusion that it's not possible.

This code has to be part of the `ActivityManager` - otherwise, we cannot synchronise it correctly. And the `ActivityManager` has to assume a very proactive role in this process (I actually think that you've been after that anyway).

I'm just about to finish a demo for a `MagicDialog` that I'd really like to be used as a base for what you have now as `ExpandDialog`. After I've finished that demo, I'll focus on the `ActivityManager`. I think that I should in general focus on the RepoMan stuff and leave the UI stuff with you for now - OTOH I'd really like to see my magic dialog as a base, so I'd ask you if you could take that code then and properly integrate it into core. Would that be okay with you?

scunz commented:
>  Could you please fix the following compiler error

Let me see, if I forgot to commit/push something. However, the right fix is to replace `MacGitver::log()` with `Log::Manager()`.

scunz commented:
I've done that change and pushed it.

antis81 committed:
use assignment operator instead of copy c'tor

antis81 committed:
restart timer when active while changing the interval

antis81 committed:
set text according to state of progress (running/finished)

TODO: error state is missing

antis81 committed:
removed obsolete method ProgressDlg::remoteMessage

antis81 commented:
> ... OTOH I'd really like to see my magic dialog as a base, so I'd ask you if you could take that code then and properly integrate it into core. Would that be okay with you?

Yeah totally fine. The current ExpandableDlg has a little problem anyways. Unlike the first implementation, the options widget is always instantiated. This currently leads to the fact, that all options are actually set.

scunz committed:
Core files for libActivities

antis81 committed:
minor cleanup in ProgressDlg

antis81 committed:
ProgressDlg: moved initializers back to c'tor

antis81 committed:
ProgressDlg: removed customized palette for log view

antis81 committed:
ProgressDlg: minor cleanup

antis81 committed:
ProgressDlg: add methods isDone() and setCanClose()

antis81 committed:
ProgressDlg: move asserts

antis81 committed:
ProgressDlg: remove invalid activities from progress updating

Activities become invalid, when they where deleted (pointer == nullptr).

antis81 committed:
add ProgressDlg::setError()

antis81 committed:
ProgressDlg: prevent accept() or reject() before activities are finished

antis81 committed:
PrgressDlg: frequently update close buttons

antis81 committed:
ProgressDlg: set step percentage to 100% when finished

antis81 committed:
ProgressDlg: mark finished and error state of an activity by coloring the progress bar

antis81 committed:
ProgressDlg: fixed erase activities

antis81 committed:
ProgressDlg: make internal progress widgets use the cells background on painting

antis81 committed:
ProgressDlg: a progress widget has no "running" state; an "active" flag better reflects what I mean here

antis81 committed:
ProgressDlg: refine the appearance of progress bars by using the activiy's result and minor cleanup

antis81 commented:
From a visual perspective, the dialog got a really nice yet simple look; some impressions taken from a test clone of Qt-Creator:

![mgv-prog-dlg-finished-one](https://cloud.githubusercontent.com/assets/440517/7430578/88922ede-f015-11e4-955d-148960a2bbc5.png)

By default, an activity will appear collapsed:
![mgv-prog-dlg-finished-all](https://cloud.githubusercontent.com/assets/440517/7430609/d32b746e-f015-11e4-8c8b-688de3fae73d.png)

Expanded - all activities finished;
![mgv-prog-dlg-finished-collapsed](https://cloud.githubusercontent.com/assets/440517/7430595/b40c0e4a-f015-11e4-8a7b-40cfeddde6a9.png)

Oops :zap: - something went wrong:
![mgv-prog-dlg-failed-activity](https://cloud.githubusercontent.com/assets/440517/7430615/e530f63e-f015-11e4-8664-9946c9f0b0e5.png)

The code design could be better. Mainly two points are holding me back to merge this:
1. :beetle: :zap: The dialog causes casual crashes and I'm not quite sure, where to fix that. @scunz Some review could help (see the timer method `ProgressDlg::updateActivities`).
2. The open points on the top :smile:

@BugSniffer FYI :eyes:

scunz commented on Libs/libMacGitverCore/Widgets/ProgressDlg.cpp:
```diff
@@ -0,0 +1,261 @@
+{
+    Activities::Iterator it = mActivities.begin();
+    while ( it != mActivities.end() ) {
+        Private::ProgressWdgt* a = *it;
```
Missing a `nullptr` check here.

antis81 replied:
No - code flow is sequential. That means any instance of `ProgressWdgt` is destroyed **only** in the destructor of `ProgressDlg`. A ProgressWdgt **must not** be destroyed before the assigned running activity has stopped (ok or with error).

IOW: Each `ProgressWdgt` **is** valid until the `ProgressDlg` is closed/destroyed.

I'll put an assert as a reminder here anyways.

scunz replied:
I think that that assumption is bad. The `ProgressDlg` ought to live forever. Hence, when the activity is finished, the progress widgets ought to be removed. However, we should not do that immediately - So maybe this needs a "dismiss" mechanism; and when doing that we might as well introduce a "cancel" mechanism (The UI ought to be the same, but the implementation will be quite different / difficult).

Anyway, if your assumption that a widget is not deleted before the dialog is true, then there's no need to store them inside `QPointer<>`'s.

scunz commented on Libs/libMacGitverCore/Widgets/ProgressDlg.cpp:
```diff
@@ -0,0 +1,261 @@
+
+        foreach (Private::ProgressWdgt* s, a->mSteps) {
+            s->progressBar->setValue(qRound(s->mPercentage));
+            qreal stepPercent = s->mPercentage / a->mSteps.size();
```
could raise a division-by-zero exception.

antis81 replied:
No. We are iterating over `a->mSteps`. Thus, `a->mSteps.size()` cannot be 0.

scunz replied:
Not in my book. However, once `libActivities` is done, you won't need to calculate that data since it has it ready for you (and with a slightly different algorithm).

antis81 commented:
@scunz Thanks for the hints concerning the crash. Unfortunately (actually fortunately :smile:) both are not relevant to the "problem". I had some uncommitted changes to GitWrap, playing with the destructors in `Git::CloneOperation` and it's base classes. After I reverted those changes, I cannot reproduce the crashes anymore - which is good actually :smile:. I'll recheck to verify, but likely there's no problem at all.

antis81 commented:
@scunz Our progress dialog is now available from a central place, but still instantiated per activity. The concept requires a single instance, where activities are added (registered) and then run in parallel. Where would you like that to be added?

When this is done, the dialog is fairly complete and usable. I plan to merge at this point. It is however not yet possible to register subactivities (we could use this to realize `git clone --recursive`).

This is only a first small step in the right direction. When it comes to visual appearance, I take Firefox and Qt-Creator as an inspirational source. There's only an icon, showing a summed progress bar of all running activities. Clicking on it (keyboard shortcut) reveals the detailed dialog. However. This is for the future .... :smile:

scunz commented:
> Where would you like that to be added?

Inside `libActivities` :)

> When this is done, the dialog is fairly complete and usable. I plan to merge at this point.

It has to be ported to `libActivities`. Sorry that I've still not finished it, but should be soonish now.

scunz commented:
Btw.: The middle step of the clone operation is not about adding objects to the git index. It's about creating an inventory for the downloaded pack file. That's why the total number is equal to the amount of downloaded objects. The total number would be equal to the number of files + directories, if this was about building the actual "git index".
The "git index" is probably created directly before the checkout (iirc libgit2 can actually only checkout indicies not trees). But I'm not sure about that.

scunz commented:
> This is only a first small step in the right direction.

I totally agree to that!!!

A few things that I'd like to be changed though:
- Can you reduce the size of the widget and remove the "finished" text once the activity is done?
- Actually: Can't we drop that line completely?

Could we try a little experiment: When this "Downloaded 500 of 1000 objects (1234mb)" text is set, could we try to show that text inside the progress bar instead of the "50%"? Would be awesome if that's possible!

I reckon that you're right now lacking a mechanism to show "details" as they come from the side-band channel in the git protocol. I can't actually see any other use of the "details" pane. But I'll implement a way to do that in `libActivities` respecting all the peculiarities of the side band transfer...

scunz commented on Libs/libMacGitverCore/Widgets/ProgressDlg.cpp:
```diff
@@ -0,0 +1,261 @@
+
+    a->mActive = false;
+    a->txtStatusInfo->setText(tr("Finished!"));
+    a->resultChanged(Private::ProgressWdgt::Result::Ok);
```
Just a side node:
`enum` doesn't introduce a name scope (C++11's `enum class` does). Thus, when you refer to `Result::Ok` some (i.e. Microsoft's) compilers complain. OTOH if you're using `enum class`, then the scope name (the enum's name) is mandatory.

``` c++
namespace X { enum       Y { Z }; } // --> X::Z
namespace X { enum class Y { Z }; } // --> X::Y::Z

// And more interesting:

enum class Foo { None, Fooish };
enum class Bar { None, Barish };

Foo f = Foo::None;
Bar b = Bar::None;

// Contrary to traditional way:

enum Foo { FooNone, FooFooish };
enum Bar { BarNone, BarBarish };
```

antis81 replied:
Yes, I actually want to use `enum class` from now on, and change the traditional way - thanks! :)

antis81 commented:
> I reckon that you're right now lacking a mechanism to show "details" as they come from the side-band channel in the git protocol. I can't actually see any other use of the "details" pane. But I'll implement a way to do that in libActivities respecting all the peculiarities of the side band transfer...

And that's exactly why I didn't implement it yet - I'd have to do it twice :).

antis81 commented:
> Could we try a little experiment: When this "Downloaded 500 of 1000 objects (1234mb)" text is set, could we try to show that text inside the progress bar instead of the "50%"? Would be awesome if that's possible!

I'll post a screenshot ...

antis81 committed:
ProgressDlg: add an assert, because currently the activity widget *must* be valid

antis81 commented:
This is how the "inlined" text looks like:

![mgv-prog-dlg-running](https://cloud.githubusercontent.com/assets/440517/7515433/3d0deabe-f4c7-11e4-98cb-965075e5eff5.png)

![mgv-prog-dlg-finished](https://cloud.githubusercontent.com/assets/440517/7515427/372dd7c6-f4c7-11e4-9c00-1473bfc28fe8.png)

Looks pretty good to me.
- If you're ok with it, I'll also completely remove the "Running ...",  "Finished!" and "Failed!" texts.
  - (Note the missing "Not started ..." text in the "Checkout" step.)
- Need a replacement word (e.g. "Packed") to correct the text "Indexed 123/234 objects".

Off-Topic, but important: The further above mentioned crash fortunately reappeared ... it is a good old threading problem. The problem code lives in the File `CloneRepositoryDlg.cpp` in the related branch `ngf/reorganize-clone-dlg`. The `rootCloneFinished()` slot is called by the worker thread (see `Git::BaseOperation::workerFinished()`). The line `delete operation;` causes the trouble now and can crash with a "double free" error:

``` bash
*** Error in `MacGitver': double free or corruption (fasttop): 0x0000000001dd1650 ***
```

Could you please revisit this and help to get it right?

antis81 committed:
ProgressDlg: inline the info text in the progress bar

antis81 committed:
ProgressDlg: use C++11 enum class

antis81 committed:
ProgressDlg: remove "running" and "finished" text

antis81 committed:
ProgressDlg: add a more descriptive "failed" text

antis81 committed:
ProgressDlg: only the activity has a finished slot

When a step progress reaches 100%, the progress bar changes to a green background.

scunz committed:
Snapshot of a mostly complete Activities interface

Still lacking lots of implementation and a good testcase.

scunz commented:
> Could you please revisit this and help to get it right?

The fix is simple:

``` bash
cd $GITWRAP
git rm -rf libGitWrap/Operations
git push --force origin development
```

Seriously, the gitWrap operations are fucked up beyond repair - and I think that this is no news, actually. That's why we're doing all this `repoman` stuff :p

So, yes. I'm already on it :)

Back to the topic:

> Looks pretty good to me.

Exactly! Couldn't agree more. This is what we need.

> Need a replacement word (e.g. "Packed") to correct the text "Indexed 123/234 objects".

No, the wording is good now. These objects are _indexed_ - they are just not added to **the git index**. "Katalogisiert" would be a better word in German, but it sounds strange in English...

scunz commented:
Actually: Your perception of which thread sends/receives the `finished()` signal seems wrong. If I read the code correctly, this ought to happen in the thread which created the operation. However, I can't see why it's going wrong at a short sight of the code.

scunz committed:
Should fix Linux build

scunz committed:
Add Service Runner into the backend

scunz committed:
Keep instance pointer to RepoMan inside itself

scunz committed:
Add: OpenRepository Service

scunz committed:
Add: RefreshRepository Service

scunz committed:
Move AutoRefresher out of RM-namespace

scunz committed:
Remove 'open' logic'

scunz committed:
Remove useless RepoMan tests

These tests actually didn't test very much, are in the wrong library and
no longer compatible with how the RepoMan will work.

We need to investigate further in how to do actually useful testing of
the RepoMan.

scunz committed:
Invoke OpenRepository service

scunz committed:
Add: refresh() method to RepoMan

scunz committed:
Move RepoMan Config Page to Config-Area

scunz committed:
Fixup: Remove unused code

scunz committed:
Step creation

scunz committed:
Fixup: Remove deadlock

scunz committed:
Actually implement state getters in Activity

scunz committed:
Fixup: "return {}" to avoid naming the constructor

scunz committed:
Public method to request opening of Progress Dialog

scunz committed:
Fixup: Inline comment (Manager Priv.hpp)

scunz committed:
Fixup to "Temporary Code" :(

scunz committed:
Turn RM::ObjType into an Enum Class

scunz committed:
Adjustments

scunz committed:
Move BasePrivate to own .cpp file

antis81 committed:
Activities: use Q_ASSERT instead of assert (fix build on linux)

antis81 committed:
RepoMan: use Q_ASSERT instead of assert (fix build on linux)

antis81 committed:
RepoMan: fix include pathes for case sensitive file systems

scunz committed:
Be explicit about namespace of Dumper class

When we'll move most of Internal to Data, the Dumper class
resides where it is. So this'll make things easier later on

scunz committed:
Move Branch & Head Private's to own cpp

antis81 committed:
RepoMan: fix missing include

antis81 committed:
Activities-Gui: add libActivitiesGui static library with copied ProgressDlg files

scunz commented:
That looks like a good starting point

antis81 committed:
Activities-Gui: draft implementation of ProgressItem classes

antis81 commented:
I just started reimplementing the dialog (and thus, this PR), switching to an item based approach - no big surprise. I already stated out, that this PR was mainly a visual preview of what the result can look like and to "get a feeling" for what we need. Nothing more, nothing less.

I'm changing the current implementation to one, based on `QStandardItemModel` (maybe QAbstractItemModel). Actually we have a simple tree of items of the same type (ProgressItem). The code in 70cc3c0 reuses your concept from `Module/References/Branches/RefsItem.hpp`. The base classes are a good candidate for `libMacGitverCoreGui` - creates a common `TreeItem` and `TreeItemObject`. In the context of this PR, the data object will refer to either a Service, Activity or Step class. As I can see now, we don't require much of a differentiation between those types. They all should have the same data fields available (displayName, curProgress, info, ...). Anyways,

The `ProgressWdgt` will be trashed and I'm going to "custom draw" the item's rect completely by the view/delegate. This is laying out a path, allowing us to go far beyond what we have now - future music. 

For now, I'll keep the dialog and finish the things under the hood first. Actually I imagine to invent a new mode (some "activities" mode with an animated icon in the left side bar). This would add some real visual pepper while creating a clean user experience the same time :star2: :smile:.

antis81 committed:
ProgressDlg: added data fields to the ProgressItem

antis81 committed:
ProgressDlg: add parent for each ProgressItem

antis81 committed:
ProgressDlg: add item delegate to draw the items

The delegate is just a template and not done right yet.
